File: RDRCTListener.m
//
//  RDRCTListener.m
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 12.07.2023.
//

#import "RDRCTListener.h"
#import "react_native_related_digital-Swift.h"

NSString *const RDRCTRegistrationEventName = @"com.relateddigital.registration";
NSString *const RDRCTNotificationResponseEventName = @"com.relateddigital.notification_response";
NSString *const RDRCTPushReceivedEventName= @"com.relateddigital.push_received";
NSString *const RDRCTOptInStatusChangedEventName = @"com.relateddigital.notification_opt_in_status";


@interface RDRCTListener()
@property(nonatomic, strong) RDRCTEventEmitter *eventEmitter;
@end

@implementation RDRCTListener

+ (RDRCTListener *)shared {
    static RDRCTListener *rdListener_ = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        rdListener_ = [[RDRCTListener alloc] init];
    });
    return rdListener_;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        self.eventEmitter = [RDRCTEventEmitter shared];
    }
    return self;
}


#pragma mark RDPushNotificationDelegate
-(void)receivedForegroundNotification:(NSDictionary *)userInfo completionHandler:(void (^)(void))completionHandler {
    id body = [RDRCTUtils eventBodyForNotificationContent:userInfo notificationIdentifier:nil];
    [self.eventEmitter sendEventWithName:RDRCTPushReceivedEventName body:body];
    completionHandler();
}

-(void)receivedBackgroundNotification:(NSDictionary *)userInfo
                    completionHandler:(void (^)(UIBackgroundFetchResult))completionHandler {

    id body = [RDRCTUtils eventBodyForNotificationContent:userInfo notificationIdentifier:nil];
    [self.eventEmitter sendEventWithName:RDRCTPushReceivedEventName body:body];
    completionHandler(UIBackgroundFetchResultNoData);
}

-(void)receivedNotificationResponse:(UNNotificationResponse *)notificationResponse
                  completionHandler:(void (^)(void))completionHandler {
    if ([notificationResponse.actionIdentifier isEqualToString:UNNotificationDismissActionIdentifier]) {
        completionHandler();
        return;
    }

    id body = [RDRCTUtils eventBodyForNotificationResponse:notificationResponse];
    [self.eventEmitter sendEventWithName:RDRCTNotificationResponseEventName body:body];
    completionHandler();
}

#pragma mark Channel Registration Events

- (void)channelRegistrationSucceeded:(NSNotification *)notification {
//    NSMutableDictionary *registrationBody = [NSMutableDictionary dictionary];

//    NSString *channelID = RD.channel.identifier;
//    NSString *deviceToken = [RD push].deviceToken;
//
//    [registrationBody setValue:channelID forKey:@"channelId"];
//    [registrationBody setValue:deviceToken forKey:@"registrationToken"];
//    [self.eventEmitter sendEventWithName:RDRCTRegistrationEventName body:registrationBody];
}


#pragma mark RDRegistrationDelegate

- (void)notificationAuthorizedSettingsDidChange:(RDAuthorizedNotificationSettings)authorizedSettings {
    NSDictionary *body = @{
        @"optIn": @(authorizedSettings != RDAuthorizedNotificationSettingsNone),
        @"authorizedNotificationSettings": [RDRCTUtils authorizedSettingsDictionary:authorizedSettings],
        @"authorizedSettings": [RDRCTUtils authorizedSettingsArray:authorizedSettings]
    };
    [self.eventEmitter sendEventWithName:RDRCTOptInStatusChangedEventName body:body];
}


@end

File: ios/RDRCTListener.swift
//
//  RDRCTListener.swift
//  react-native-related-digital
//
//  Created by Egemen Gülkılık on 14.07.2023.
//


import Foundation
import RelatedDigitalIOS
import UserNotifications

fileprivate typealias NativeRD = RelatedDigitalIOS.RelatedDigital
fileprivate typealias UIA = UIApplication
fileprivate typealias NC = NotificationCenter
fileprivate typealias UNUNC = UNUserNotificationCenter


@objc(RDPushNotificationDelegate)
public protocol PushNotificationDelegate: NSObjectProtocol {
    @objc
    optional func receivedForegroundNotification(_ userInfo: [AnyHashable : Any], completionHandler: @escaping () -> Void)
#if !os(watchOS)
    @objc
    optional func receivedBackgroundNotification(_ userInfo: [AnyHashable : Any], completionHandler: @escaping (UIBackgroundFetchResult) -> Void)
#else
    @objc
    optional func receivedBackgroundNotification(_ userInfo: [AnyHashable : Any], completionHandler: @escaping (WKBackgroundFetchResult) -> Void)
#endif
#if !os(tvOS)
    @objc
    optional func receivedNotificationResponse(_ notificationResponse: UNNotificationResponse, completionHandler: @escaping () -> Void)
#endif
    @objc(extendPresentationOptions:notification:)
    optional func extend(_ options: UNNotificationPresentationOptions, notification: UNNotification) -> UNNotificationPresentationOptions
}


@objc(RDRegistrationDelegate)
public protocol RegistrationDelegate: NSObjectProtocol {
#if !os(tvOS)
    @objc
    optional func notificationRegistrationFinished(
        withAuthorizedSettings authorizedSettings: RDAuthorizedNotificationSettings,
        categories: Set<UNNotificationCategory>,
        status: RDAuthorizationStatus
    )
#endif
    @objc
    optional func notificationRegistrationFinished(
        withAuthorizedSettings authorizedSettings: RDAuthorizedNotificationSettings,
        status: RDAuthorizationStatus
    )
    @objc optional func notificationAuthorizedSettingsDidChange(_ authorizedSettings: RDAuthorizedNotificationSettings)
    
    @objc optional func apnsRegistrationSucceeded(withDeviceToken deviceToken: Data)

    
    @objc optional func apnsRegistrationFailedWithError(_ error: Error)
}




@objc public class RDRCTListener: NSObject
//, RDPushNotificationDelegate, RDRegistrationDelegate, RDDeepLinkDelegate, RDMessageCenterDisplayDelegate
{

    @objc public static let shared = RDRCTListener()

    let eventEmitter = RDRCTEventEmitter.shared

    private override init() {
//        super.init()
//
//        NC.default.addObserver(self, selector: #selector(inboxUpdated), name: NSNotification.Name(rawValue: RDInboxMessageListUpdatedNotification), object: nil)
//
//        NC.default.addObserver(self, selector: #selector(channelRegistrationSucceeded), name: NSNotification.Name(rawValue: RDChannel.channelUpdatedEvent), object: nil)
    }

    // MARK: RDDeepLinkDelegate

    func receivedDeepLink(_ deepLink: URL, completionHandler: @escaping () -> Void) {
//        let body = ["deepLink" : deepLink.absoluteString]
//        eventEmitter().sendEvent(withName: RDRCTDeepLinkEventName, body: body)
//        completionHandler()
    }

    // MARK: RDPushDelegate

    func receivedForegroundNotification(_ userInfo: [AnyHashable : Any], completionHandler: @escaping () -> Void) {
//        let body = RDRCTUtils.eventBody(forNotificationContent: userInfo, notificationIdentifier: nil)
//        eventEmitter.sendEvent(withName: RDRCTPushReceivedEventName, body: body)
//        completionHandler()
    }

    func receivedBackgroundNotification(_ userInfo: [AnyHashable : Any], completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
//        let body = RDRCTUtils.eventBody(forNotificationContent: userInfo, notificationIdentifier: nil)
//        eventEmitter.sendEvent(withName: RDRCTPushReceivedEventName, body: body)
//        completionHandler(.noData)
    }

    func receivedNotificationResponse(_ notificationResponse: UNNotificationResponse, completionHandler: @escaping () -> Void) {
        if notificationResponse.actionIdentifier == UNNotificationDismissActionIdentifier {
            completionHandler()
            return
        }

//        let body = RDRCTUtils.eventBody(forNotificationResponse: notificationResponse)
//        eventEmitter.sendEvent(withName: RDRCTNotificationResponseEventName, body: body)
//        completionHandler()
    }

    // MARK: Channel Registration Events

    @objc func channelRegistrationSucceeded(_ notification: Notification) {
        //var registrationBody = [String: Any]()

        //let channelID = RelatedDigital.channel.identifier
        //let deviceToken = RelatedDigital.push().deviceToken

        //registrationBody["channelId"] = channelID
        //registrationBody["registrationToken"] = deviceToken
        //eventEmitter.sendEvent(withName: RDRCTRegistrationEventName, body: registrationBody)
    }

    // MARK: RDRegistrationDelegate

    func notificationAuthorizedSettingsDidChange(_ authorizedSettings: RDAuthorizedNotificationSettings) {
//        let body = [
//            "optIn": authorizedSettings != .none,
//            "authorizedNotificationSettings": RDRCTUtils.authorizedSettingsDictionary(authorizedSettings),
//            "authorizedSettings": RDRCTUtils.authorizedSettingsArray(authorizedSettings)
//        ]
//        eventEmitter.sendEvent(withName: RDRCTOptInStatusChangedEventName, body: body)
    }

    // MARK: Message Center

    func displayMessageCenter(forMessageID messageID: String, animated: Bool) {
//        if RDRCTStorage.autoLaunchMessageCenter {
//            RDMessageCenter.shared().defaultUI.displayMessageCenter(forMessageID: messageID, animated: animated)
//        } else {
//            eventEmitter.sendEvent(withName: RDRCTShowInboxEventName, body: ["messageId": messageID])
//        }
    }

    func dismissMessageCenter(animated: Bool) {
//        if RDRCTStorage.autoLaunchMessageCenter {
//            RDMessageCenter.shared().defaultUI.dismissMessageCenter(animated: animated)
//        }
    }

    @objc func inboxUpdated() {
//        let body = [
//            "messageUnreadCount": RDMessageCenter.shared().messageList.unreadCount,
//            "messageCount": RDMessageCenter.shared().messageList.messageCount
//        ]
//
//        eventEmitter.sendEvent(withName: RDRCTInboxUpdatedEventName, body: body)
    }
}
